python/ray/util/spark/cluster_init.py

```diff
@@ -489,35 +543,85 @@ def _setup_ray_cluster(
-        "--block",
-        "--head",
-        f"--node-ip-address={ray_head_ip}",
-        f"--port={ray_head_port}",
-        # disallow ray tasks with cpu/gpu requirements from being scheduled on the head
-        # node.
-        f"--num-cpus={num_cpus_head_node}",
-        f"--num-gpus={num_gpus_head_node}",
-        f"--memory={heap_memory_head_node}",
-        f"--object-store-memory={object_store_memory_head_node}",
-        *dashboard_options,
-        *_convert_ray_node_options(head_node_options),
-    ]
-
-    _logger.info(f"Starting Ray head, command: {' '.join(ray_head_node_cmd)}")
-
-    ray_head_proc, tail_output_deque = exec_cmd(
-        ray_head_node_cmd,
-        synchronous=False,
-        extra_env={
-            RAY_ON_SPARK_COLLECT_LOG_TO_PATH: collect_log_to_path or "",
-            RAY_ON_SPARK_START_RAY_PARENT_PID: str(os.getpid()),
-        },
-    )
+    if autoscale:
+        from ray.autoscaler._private.spark.spark_job_server import (
+            _start_spark_job_server,
+        )
+
+        spark_job_server = _start_spark_job_server(
+            ray_head_ip, spark_job_server_port, spark
+        )
+        autoscaling_cluster = AutoscalingCluster(
+            head_resources={
+                "CPU": num_cpus_head_node,
+                "GPU": num_gpus_head_node,
+                "memory": heap_memory_head_node,
+                "object_store_memory": object_store_memory_head_node,
+            },
+            worker_node_types={
+                "ray.worker": {
+                    "resources": {
+                        "CPU": num_cpus_worker_node,
+                        "GPU": num_gpus_worker_node,
+                        "memory": heap_memory_worker_node,
+                        "object_store_memory": object_store_memory_worker_node,
+                    },
+                    "node_config": {},
+                    "min_workers": 0,
+                    "max_workers": num_worker_nodes,
+                },
+            },
+            extra_provider_config={
+                "ray_head_ip": ray_head_ip,
+                "ray_head_port": ray_head_port,
+                "cluster_unique_id": cluster_unique_id,
+                "using_stage_scheduling": using_stage_scheduling,
+                "ray_temp_dir": ray_temp_dir,
+                "worker_node_options": worker_node_options,
+                "collect_log_to_path": collect_log_to_path,
+                "spark_job_server_port": spark_job_server_port,
+            },
+            upscaling_speed=autoscale_upscaling_speed,
+            idle_timeout_minutes=autoscale_idle_timeout_minutes,
+        )
+        ray_head_proc, tail_output_deque = autoscaling_cluster.start(
+            ray_head_ip,
+            ray_head_port,
+            ray_temp_dir,
+            dashboard_options,
+            head_node_options,
+            collect_log_to_path,
+        )
+        ray_head_node_cmd = autoscaling_cluster.ray_head_node_cmd
+    else:
```

Comment:
Similar to AutoscalingCluster, can we also create a `StaticCluster` class to encapsulate the logic of starting head and worker nodes of a static cluster?

Reply:
This is code refactoring work, I suggest we do it in follow-up PR.

Reply:
ok, let's refactor in a follow-up PR. Also this file is very big now. We should also split it into multiple smaller files during refactoring.